framework/src/service/ServiceReferenceBasePrivate.cpp

```diff
@@ -43,34 +43,29 @@ namespace cppmicroservices
 
     using ThreadMarksMapType = std::unordered_map<BundlePrivate*, std::unordered_set<ServiceRegistrationBasePrivate*>>;
 
-    ServiceReferenceBasePrivate::ServiceReferenceBasePrivate(ServiceRegistrationBasePrivate* reg)
-        : ref(1)
-        , registration(reg)
+    ServiceReferenceBasePrivate::ServiceReferenceBasePrivate(std::weak_ptr<ServiceRegistrationBasePrivate> reg)
+        : registration(reg)
     {
-        if (registration)
-            ++registration->ref;
-    }
-
-    ServiceReferenceBasePrivate::~ServiceReferenceBasePrivate()
-    {
-        if (registration && !--registration->ref)
+        if (!reg.expired())
```

Comment:
Can registration be a nullptr? If so, should be checked before using. The old code checked before using which implies it could be a nullptr